### PR TITLE
[SHEL-80] User Role Based Validation and Product routes

### PR DESCRIPTION
**Pull request by wit03**
- Create all CRUD product routes.
- Create one type conversion function (from entity to dto)
- Remove false foreign key from gorm migration process.
- Cleanup some unused code.

### CODE REVIEW

File: internal/product/product.service.go [hunk 1]
```diff
@@ -22,9 +29,176 @@ func NewProductService(productUsecase ProductUsecase) ProductService {
...
+		Category:       req.Category,
+		Subcategory:    req.Subcategory,
+		Description:    req.Description,
+		IsVerified:     req.IsVerified,
+		Price:          req.Price,
+		ImageURL:       req.ImageURL,
+		VaccineBookURL: req.VaccineBookURL,
+	}
+
+	productRes, err := s.productUsecase.InsertProduct(ctx.Context(), &product)
```
Comment (TGRZiminiar) [inline internal/product/product.service.go:60]: Should use ctx.UserContext instead not just here by should use everywhere when want to pass down the context

File: internal/product/product.service.go [hunk 2]
```diff
@@ -22,9 +29,176 @@ func NewProductService(productUsecase ProductUsecase) ProductService {
...
+
+	if err != nil {
+		return ctx.Status(fiber.StatusInternalServerError).JSON(dto.HttpResponse{
+			Error:   "Failed to insert product",
+			Success: false,
+		})
+	}
+
+	return ctx.Status(fiber.StatusOK).JSON(dto.HttpResponse{
```
Comment (TGRZiminiar) [inline internal/product/product.service.go:69]: Should use fiber.StatusCreated instead

**Review by TGRZiminiar — CHANGES_REQUESTED** [review]

**Comment by wit03** [comment]
@TGRZiminiar Done

File: internal/product/product.service.go [hunk 3]
```diff
@@ -22,9 +29,176 @@ func NewProductService(productUsecase ProductUsecase) ProductService {
...
+	}
+
+	return ctx.Status(fiber.StatusCreated).JSON(dto.HttpResponse{
+		Result:  productRes,
+		Success: true,
+	})
+}
+
+func (s *productService) GetProduct(ctx *fiber.Ctx) error {
+	products, err := s.productUsecase.GetProduct(ctx.Context())
```
Comment (TGRZiminiar) [inline internal/product/product.service.go:76]: use Usercontext

File: internal/product/product.service.go [hunk 4]
```diff
@@ -22,9 +29,176 @@ func NewProductService(productUsecase ProductUsecase) ProductService {
...
+	idParam := ctx.Params("id")
+	id, err := strconv.ParseUint(idParam, 10, 64)
+	if err != nil {
+		return ctx.Status(fiber.StatusBadRequest).JSON(dto.HttpResponse{
+			Error:   "Invalid ID",
+			Success: false,
+		})
+	}
+
+	product, err := s.productUsecase.GetProductByID(ctx.Context(), uint(id))
```
Comment (TGRZiminiar) [inline internal/product/product.service.go:99]: use Usercontext

File: internal/product/product.service.go [hunk 5]
```diff
@@ -22,9 +29,176 @@ func NewProductService(productUsecase ProductUsecase) ProductService {
...
+		Category:       req.Category,
+		Subcategory:    req.Subcategory,
+		Description:    req.Description,
+		IsVerified:     req.IsVerified,
+		Price:          req.Price,
+		ImageURL:       req.ImageURL,
+		VaccineBookURL: req.VaccineBookURL,
+	}
+
+	err = s.productUsecase.UpdateProduct(ctx.Context(), &updateProduct, uint(id), userId)
```
Comment (TGRZiminiar) [inline internal/product/product.service.go:151]: use Usercontext

File: internal/product/product.service.go [hunk 6]
```diff
@@ -22,9 +29,176 @@ func NewProductService(productUsecase ProductUsecase) ProductService {
...
+	if err != nil {
+		return ctx.Status(fiber.StatusInternalServerError).JSON(dto.HttpResponse{
+			Error:   err.Error(),
+			Success: false,
+		})
+	}
+
+	// Get updated product and response
+	product, err := s.productUsecase.GetProductByID(ctx.Context(), uint(id))
```
Comment (TGRZiminiar) [inline internal/product/product.service.go:160]: use Usercontext

File: internal/product/product.service.go [hunk 7]
```diff
@@ -22,9 +29,176 @@ func NewProductService(productUsecase ProductUsecase) ProductService {
...
+
+	userId, err := jwt.GetIDFromToken(ctx.Cookies("access_token"))
+	if err != nil {
+		return ctx.Status(fiber.StatusInternalServerError).JSON(dto.HttpResponse{
+			Error:   "Cannot get UserID from access_token context: " + err.Error(),
+			Success: false,
+		})
+	}
+
+	err = s.productUsecase.DeleteProduct(ctx.Context(), userId, uint(id))
```
Comment (TGRZiminiar) [inline internal/product/product.service.go:192]: use Usercontext

**Review by TGRZiminiar — CHANGES_REQUESTED** [review]

**Pull request merged**